metaflow/runtime.py

```diff
@@ -666,7 +666,7 @@ def _queue_task_join(self, task, next_steps):
                 # Control task was successful.
                 # Additionally check the state of (sibling) mapper tasks as well
                 # (for the sake of resume) before queueing join task.
-                num_splits = self._control_num_splits[control_path]
```

Comment:
can you describe the scenario to reproduce it? (I think line 671-672 we are checking the siblings, so we need to get the num_splits from the parent/control path, is this not the expected logic?

Reply:
Try running this and then resuming it : 

```python
from metaflow import FlowSpec, step, parallel, current

class SimpleParallelFlow(FlowSpec):
    
    @step
    def start(self):
        # Create some data to process in parallel
        self.data = list(range(10))
        # Split into 4 parallel tasks
        self.next(self.process_data, num_parallel=2)
    
    @parallel
    @step
    def process_data(self):
        # Print which parallel node we're on
        print(f"Processing on node {current.parallel.node_index}")
        
        # Simulate some work
        import time
        time.sleep(2)
        
        # Store result specific to this parallel task
        self.result = f"Result from node {current.parallel.node_index}"
        
        self.next(self.join)
    
    @step
    def join(self, inputs):
        # Collect results from all parallel tasks
        raise Exception("test")
        self.results = [input.result for input in inputs]
        print("Combined results:", self.results)
        
        self.next(self.end)
    
    @step
    def end(self):
        print("Flow completed with results:", self.results)

if __name__ == "__main__":
    SimpleParallelFlow()
```

I don't know if this is the correct solution but here is some diagnosis for you. 

The error I am getting looks something like for the above flow on calling resume post run: 

```
Metaflow 2.12.30.post6-gitfbc9f1f-dirty+obcheckpoint(0.1.4);ob(v1) executing SimpleParallelFlow for user:valay@outerbounds.co
Validating your flow...
    The graph looks good!
Running pylint...
    Pylint is happy!
2024-12-04 12:57:09.492 Gathering required information to resume run (this may take a bit of time)...
2024-12-04 12:57:15.980 Workflow starting (run-id 7316), see it in the UI at https://ui.dev-valay.outerbounds.xyz/p/default/SimpleParallelFlow/7316
2024-12-04 12:57:16.340 Cloning SimpleParallelFlow/7307
Skipping cloning mapper task 7307/start/46646
2024-12-04 12:57:16.540 [7316/process_data/46678] Cloning previously run task 7307/process_data/46647
2024-12-04 12:57:16.987 SimpleParallelFlow/7307 cloned!
self._control_num_splits {'7316/process_data/46678': 2}
2024-12-04 12:57:17.064 Workflow failed.
2024-12-04 12:57:17.064 Terminating 0 active tasks...
2024-12-04 12:57:17.064 Flushing logs...
    Internal error
Traceback (most recent call last):
  File "/home/ubuntu/metaflow/metaflow/cli.py", line 1167, in main
    start(auto_envvar_prefix="METAFLOW", obj=state)
  File "/home/ubuntu/metaflow/metaflow/tracing/tracing_modules.py", line 106, in wrapper_func
    return func(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/_vendor/click/core.py", line 829, in __call__
    return self.main(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/_vendor/click/core.py", line 782, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/_vendor/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/_vendor/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/_vendor/click/core.py", line 610, in invoke
    return callback(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/cli.py", line 586, in wrapper
    return func(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/_vendor/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/metaflow/metaflow/cli.py", line 738, in resume
    runtime.execute()
  File "/home/ubuntu/metaflow/metaflow/runtime.py", line 437, in execute
    self._queue_tasks(finished_tasks)
  File "/home/ubuntu/metaflow/metaflow/runtime.py", line 822, in _queue_tasks
    self._queue_task_join(task, next_steps)
  File "/home/ubuntu/metaflow/metaflow/runtime.py", line 674, in _queue_task_join
    num_splits = self._control_num_splits[control_path]
                 ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
KeyError: '7307/process_data/46647'

```

Its failing at this line because even if the [control task's num splits](https://github.com/Netflix/metaflow/blob/c72c93f8159266110e45d7782900651636f29d99/metaflow/runtime.py#L638) are set, the value of the `control_path` is extracted from `self._finished`. This yields a `control_path` which points to a pathspec of the origin run during a resume, The value of the key set in the `self._control_num_splits` is of the current run's task but the `control_path` being the origin run's task, it results in the key error.

Reply:
Another minimal example. I noticed resuming this run on fresh Metaflow install fails. 

```python
from metaflow import FlowSpec, step, kubernetes, parallel

N_NODES = 2


class HelloParallel(FlowSpec):
    @step
    def start(self):
        self.next(self.f, num_parallel=N_NODES)

    @parallel
    @kubernetes
    @step
    def f(self):
        self.next(self.join)

    @step
    def join(self, inputs):
        # raise ValueError('Oh no!') # uncomment this on run, comment on resume
        self.next(self.make_failure)

    @step
    def make_failure(self):
        self.next(self.end)

    @step
    def end(self):
        pass


if __name__ == "__main__":
    HelloParallel()
```